types/geojson/index.d.ts

```diff
@@ -38,7 +38,7 @@ export type BBox = [number, number, number, number] | [number, number, number, n
  * The previous GeoJSON specification allowed more elements (e.g., which could be used to represent M values),
  * but the current specification only allows X, Y, and (optionally) Z to be defined.
  */
-export type Position = number[]; // [number, number] | [number, number, number];
+export type Position = [number, number, ...number[]]; // [number, number] | [number, number, number];
```

Comment:
If the comment above is to be believed, then shouldn't this be `[number, number, number?]`? (i.e. exactly two or three numbers -- equivalent to the commented-out type.)

This change has been proposed and rejected before as being too drastic, see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/21590#discussion_r151582350 and cc @DenisCarriere and @JeffJacobson. I tend to agree -- while the proposed type is more precise, this will be almost certainly be a breaking change for all users of these types, and I'm not sure the added safety is really worth the cost. I patched this into my main work project just now, for example, and got 35 type errors.

Reply:
@danvk `[number, number, number?]` would accept `[0, 0, undefined]`

Reply:
Don't wait for my approval; whatever the rest of you decide is fine. It's been a very long time since I've needed to use the GeoJSON type definition (which was several versions of TypeScript ago), so I don't know what is best at the moment.

In retrospect, I regret changing the version number major to 7946 to match the RFC number. In [semantic versioning](https://semver.org/), if there's a breaking change you're supposed to change the major portion of the version number, but now if you do that it won't match the RFC number.

Is there some way in TypeScript to support the both the more restrictive and less restrictive types, but mark the less restrictive one as being deprecated? So a user using more than three coordinates in a position would merely see a warning, but their code wouldn't break?

Alternatively, could an additional alternative type definition be provided (called something like "geojson-strict") with the more restrictive position definition?

Below is an excerpt from [RFC 7946 Section 3.1.1](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.1), which states the restriction to either two or three elements for easier reference. (I bolded some of the text which specifies the number of allowed elements.)

>    A position is an array of numbers.  **There MUST be two or more
>    elements.**  The first two elements are longitude and latitude, or
>    easting and northing, precisely in that order and using decimal
>    numbers.  Altitude or elevation MAY be included as an optional third
>    element.
> 
>    **Implementations SHOULD NOT extend positions beyond three elements**
>    because the semantics of extra elements are unspecified and
>    ambiguous.  Historically, some implementations have used a fourth
>    element to carry a linear referencing measure (sometimes denoted as
>    "M") or a numerical timestamp, but in most situations a parser will
>    not be able to properly interpret these values.  The interpretation
>    and meaning of additional elements is beyond the scope of this
>    specification, and additional elements MAY be ignored by parsers.

Reply:
> @danvk `[number, number, number?]` would accept `[0, 0, undefined]`

Would it also support an array of only two numbers?

Reply:
@JeffJacobson yes, but my point is that `[0, 0, undefined]` should **not** be a valid `Position`.

Reply:
> @JeffJacobson yes, but my point is that `[0, 0, undefined]` should **not** be a valid `Position`.

Thanks for clarifying. I misunderstood.

Reply:
If you set `exactOptionalPropertyTypes` then `[0, 0, undefined]` is not assignable to that type ([playground][1]):

```ts
type Position = [number, number, number?];

const p: Position = [1, 2, undefined];
//    ~ Type 'undefined' is not assignable to type 'number'.
```

[1]: https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/C4TwDgpgBACg9gZwJbCXAdlAvFA2ugVwFsAjCAJwBopDSLrazyB+AXQG4AoTgYwwWBQwALliIUaTDlwBGagCZqBdABMIAMyToIKjpwD0+qMagA-KABVw0AOTK1m7SptQkCGnEEBDBMgDm6F4kADbQwHBQoJBQNowUNgB0nEA

Reply:
That's good to know. However, considering that most people don't have `exactOptionalPropertyTypes` enabled, I'd still choose `[number, number] | [number, number, number]` over `[number, number, number?]`

Reply:
In the proposal i made, it is not as strict as [number, number] | [number, number, number]
It just enforce to have at least 2 number values in the array, as it should be for a coordinate position. Also look at the official Json schema definition from geojson.

Reply:
@apiel is this change something you personally need? Or is the motivation purely to be more compliant with the spec?

Reply:
Originally, I needed it when I was using https://www.npmjs.com/package/typescript-json-schema to generate JSON schema for my swagger OpenAPI documentation. Since the types were not consistent, I ended up to download the JSON schema from  https://geojson.org/schema/GeoJSON.json and directly linked them to my doc (so I don't rely on TypeScript anymore to represent my API).

However, TypeScript is meant to be type safe. I still think that a position should have at least 2 values. So would be great to have it right in the types, especially since we have a clear type definition on geojson.org